content/post/postgres/extension-preloading.md

```diff
@@ -0,0 +1,170 @@
+
+Normal Extensions
+-----------------
+
+If your extension includes no loadable modules, congratulations! You don't
+have to worry about this question at all.
+
+If your extension's shared library provides functionality only via functions
+called from SQL, you also don't need to worry about this preloading. Custom
+types, operators, and functions generally follow this pattern. The DDL that
+creates the SQL object, such as [`CREATE FUNCTION`], uses the `AS 'obj_file',
+'link_symbol'` syntax to tell that PostgreSQL what library to load when it's
+needed.
+
+Hook Extensions
+---------------
+
+If your extension's shared library makes calls to PostgreSQL without
+PostgreSQL first calling it, then the library must be loaded before it's used.
+This is typically the case for libraries that modifies the server's behavior
+through "hooks" rather than providing a set of functions.
+
+### Session Preloading
+
+If your extension is intended for debugging or performance-measurement, it
+likely doesn't need to be preloaded for every connection. In this case, a DBA
+might allow specific users to load it by either:
+
+*   Adding it to the [`session_preload_libraries`] variable for the user via
+    [`ALTER ROLE`], so it loads for every connection for that user:
+
+    ```sql
+    ALTER ROLE role_name
+      SET session_preload_libraries TO '$libdir/mylib';
+    ```
+
+*   Granting the user role the ability to set `session_preload_libraries`
+    which would allow them to use it (and any other shared library) in
+    `PGOPTIONS`:
+
+    ```sql
+    GRANT SET ON PARAMETER session_preload_libraries
+       TO role_name;
+    ```
+
+As an extension author, you don't need to do any special configuration, as
+long as your module is installed in the usual location via the [`MODULES`]
+`Makefile` variable. Still, it will be useful to document these options so
+that DBAs quickly see how to set things up for the users who need them.
+
+### Local Preloading
+
+As a special case, a DBA might want to make your debugging or
+performance-measurement extension available to any user who needs it, even
+unprivileged users. All they need to do is move it from `$libdir` to
+`$libdir/plugins`.
+
+Then any Postgres user can load it via the [`LOAD`] command or include it in
+their [`local_preload_libraries`] configuration via either [`PGOPTIONS`] or,
+for every connection, via `ALTER ROLE SET`:
+
+```sql
+ALTER ROLE role_name
+  SET local_preload_libraries TO '$libdir/plugins/mylib';
+```
+
+As an extension author, you don't need to do any special configuration; there
+is no `Makefile` variable to install it in the the `$libdir/plugins`
+directory. But it might be handy for DBAs to document this option *in addition
+to* the [session preloading](#session-preloading) options. But emphasize that
+it should be used if and *only if* they want to allow any and all of their
+users to load your extension library without any further intervention.
```

Comment:
The docs make a point that extensions need to be coded to be run by non-superusers, but doesn't elaborate on that point at all.  Maybe do some elaboration here?  Like, if you add a function to the server than can only be executed by a superuser then the non-privileged user is just going to see an error while or after loading the extension.

Reply:
In efda14c, I added this bit:

> As long as no function or operation in your extension *requires* superuser access and doesn't provide SQL objects that map to `$libdir/mylib`, things should work as before.